python/task/workflow.py

```diff
@@ -0,0 +1,703 @@
+      writer.attach_output("output", output[shard])
+      writer_tasks.append(writer)
+
+    # Connect producer(s) to writer task(s).
+    if isinstance(input, list) and len(input) == 1: input = input[0]
+    if fanout == 1: writer_tasks = writer_tasks[0]
+    self.connect(input, writer_tasks)
+
+    return output
+
+  def collect(self, *args):
+    """Return list of channels that collects the input from all the arguments.
+    The arguments can be channels, resources, or lists of channels or
+    resources."""
+    channels = []
+    for arg in args:
+      if isinstance(arg, Channel):
+        channels.append(arg)
+      elif isinstance(arg, Resource):
+        channels.append(self.read(arg))
+      elif isinstance(arg, list):
+        for elem in arg:
+          if isinstance(elem, Channel):
+            channels.append(elem)
+          elif isinstance(elem, Resource):
+            channels.append(self.read(elem))
+          else:
+            raise Exception("illegal element")
+        pass
+      else:
+        raise Exception("illegal argument")
+    return channels if len(channels) > 1 else channels[0]
+
+  def parallel(self, input, threads=5, queue=None, name=None):
+    """Parallelize input messages over thread worker pool."""
+    workers = self.task("workers", name=name)
+    workers.add_param("worker_threads", threads)
+    if queue != None: workers.add_param("queue_size", queue)
+    self.connect(input, workers)
+    return self.channel(workers, format=format_of(input))
+
+  def map(self, input, type=None, format=None, params=None, name=None):
+    """Map input through processor."""
+    # Use input format if no format specified.
+    if format == None: format = format_of(input).as_message()
+
+    # Create mapper.
+    if type != None:
+      mapper = self.task(type, name=name)
+      mapper.add_params(params)
+      reader = self.read(input)
+      self.connect(reader, mapper)
+      output = self.channel(mapper, format=format)
+    else:
+      output = self.read(input)
+
+    return output
+
+  def shuffle(self, input, shards):
+    """Shard and sort the input messages."""
+    # Create sharder and connect input.
+    sharder = self.task("sharder")
+    self.connect(input, sharder)
+    pipes = self.channel(sharder, shards=shards, format=format_of(input))
+
+    # Pipe outputs from sharder to sorters.
+    if shards != None:
+      sorters = []
+      for i in xrange(shards):
+        sorter = self.task("sorter", shard=Shard(i, shards))
+        self.connect(pipes[i], sorter)
+        sorters.append(sorter)
+    else:
+      sorters = self.task("sorter")
+      self.connect(pipes, sorters)
+
+    # Return output channel from sorters.
+    outputs = self.channel(sorters, format=format_of(input))
+    return outputs
+
+  def reduce(self, input, output, type=None, params=None, name=None):
+    """Reduce input and write reduced output."""
+    if type == None:
+      # No reducer (i.e. i.e. identity reducer), just write input.
```

Comment:
typo: duplicate "i.e."

Reply:
Fixed. However, the comment is slightly misleading since we don't actually use an identity reducer in this case. The task system is more flexible that a standard mapreduce, where you are forced to provide both a mapper and a reducer.